### ISSUE TITLE
Issue with Model Inference: Same Class Number Predicted for All Images

### ISSUE THREAD

**Issue by mahrokh3409** [nasaharvest/presto#40]
Hi @gabrieltseng,

I used the CropHarvestMultiClassValidation class from presto/eval/cropharvest_eval.py for finetuning. After training with the dataset provided by this class, I saved the finetuned model and attempted to perform inference. However, the model predicts the same class number for all images.

_model_save_path = "finetuned_model.pth"  
fine_tuned_model = torch.load(model_save_path)
fine_tuned_model.to(device)
fine_tuned_model.eval()

with torch.no_grad():
    outputs = fine_tuned_model(
        arrays.float().to(device),
        dynamic_world=dynamic_world.long().to(device),
        mask=masks.bool().to(device),
        latlons=latlons.float().to(device),
        month=month.to(device)
    ).cpu().numpy()
    preds = np.argmax(outputs, axis=-1)
    print(preds.shape)
    print(preds)_

The issue is that preds contains the same class number for every image that i pass to the model. Could you help me identify what might be going wrong?

I really appreciate your help and prompt response.

Kid Regrads,
Mahrokh

**Comment by gabrieltseng** [comment]
Hi @mahrokh3409 ,

Could you share your finetuning code as well, and the model weights? 

Also, what data are you performing inference against?

**Comment by mahrokh3409** [comment]
Hi @gabrieltseng 
Thank you so much for your prompt response.

i added the below code to cropharvest_eval.py

_if __name__ == "__main__":
    val_ratio = 0.2  
    n_per_class = 200 
    ignore_dynamic_world = True  
    seed = 42 

    class_instance = CropHarvestMultiClassValidation(
        val_ratio=val_ratio,
        n_per_class=n_per_class,
        ignore_dynamic_world=ignore_dynamic_world,
        seed=seed
    )

    pretrained_model = presto.Presto.load_pretrained()
    pretrained_model.to(device)

    model_save_path ='finetuded_model/finetuned_model.pth' 
    fine_tuned_model = class_instance.finetune(pretrained_model)
    torch.save(fine_tuned_model, model_save_path)
    print(f"Finetuned Model is saved to {model_save_path}")

    evaluation_results = class_instance.evaluate(fine_tuned_model)
    print("Main Evaluation Results:")
    print(evaluation_results)_
    
  I ran the script and got the result for fine-tuning and evaluation.

Then, for inferencing the finetuned model,
I used the below code in a separate script:

_month=2
month = torch.tensor([month]*arrays.shape[0]).long()
#print(month.shape) #torch.Size([476])


device = torch.device("cuda" if torch.cuda.is_available() else "cpu")

model_save_path = "inetuded_model/finetuned_model.pth"  
fine_tuned_model = torch.load(model_save_path)
print(type(fine_tuned_model))

fine_tuned_model.to(device)
fine_tuned_model.eval()

with torch.no_grad():   
      outputs= fine_tuned_model(arrays.float().to(device), dynamic_world=dynamic_world.long().to(device),mask=masks.bool().to(device), latlons=latlons.float().to(device), month=month.to(device)).cpu().numpy()
      preds= np.argmax(outputs, axis=-1)   
      print(preds.shape)  #torch.Size([476, 10])
      print(preds)

for inference, I used eo_data images with dimension (206, 17,22) as an example.
"I used presto.construct_single_presto_input for x, mask, dynamic_word"
I appreciate your help.

Please let me know if you need any further information.

Kind Regards,
Mahrokh

**Comment by mahrokh3409** [comment]
[presto_inference.txt](https://github.com/user-attachments/files/16729157/presto_inference.txt)

**Comment by mahrokh3409** [comment]
<class 'presto.presto.PrestoFineTuningModel'>
PrestoFineTuningModel(
  (encoder): Encoder(
    (eo_patch_embed): ModuleDict(
      (S1): Linear(in_features=2, out_features=128, bias=True)
      (S2_RGB): Linear(in_features=3, out_features=128, bias=True)
      (S2_Red_Edge): Linear(in_features=3, out_features=128, bias=True)
      (S2_NIR_10m): Linear(in_features=1, out_features=128, bias=True)
      (S2_NIR_20m): Linear(in_features=1, out_features=128, bias=True)
      (S2_SWIR): Linear(in_features=2, out_features=128, bias=True)
      (ERA5): Linear(in_features=2, out_features=128, bias=True)
      (SRTM): Linear(in_features=2, out_features=128, bias=True)
      (NDVI): Linear(in_features=1, out_features=128, bias=True)
    )
    (dw_embed): Embedding(10, 128)
    (latlon_embed): Linear(in_features=3, out_features=128, bias=True)
    (blocks): ModuleList(
      (0-1): 2 x Block(
        (norm1): LayerNorm((128,), eps=1e-05, elementwise_affine=True)
        (attn): Attention(
          (qkv): Linear(in_features=128, out_features=384, bias=True)
          (q_norm): Identity()
          (k_norm): Identity()
          (attn_drop): Dropout(p=0.0, inplace=False)
          (proj): Linear(in_features=128, out_features=128, bias=True)
          (proj_drop): Dropout(p=0.0, inplace=False)
        )
        (ls1): Identity()
        (norm2): LayerNorm((128,), eps=1e-05, elementwise_affine=True)
        (mlp): Mlp(
          (fc1): Linear(in_features=128, out_features=512, bias=True)
          (act): GELU(approximate='none')
          (drop1): Dropout(p=0.0, inplace=False)
          (fc2): Linear(in_features=512, out_features=128, bias=True)
          (drop2): Dropout(p=0.0, inplace=False)
        )
        (ls2): Identity()
      )
    )
    (norm): LayerNorm((128,), eps=1e-05, elementwise_affine=True)
    (month_embed): Embedding(12, 32)
    (channel_embed): Embedding(10, 32)
  )
  (head): FinetuningHead(
    (linear): Linear(in_features=128, out_features=10, bias=True)
  )
)
encoder.pos_embed torch.Size([1, 24, 64])
encoder.eo_patch_embed.S1.weight torch.Size([128, 2])
encoder.eo_patch_embed.S1.bias torch.Size([128])
encoder.eo_patch_embed.S2_RGB.weight torch.Size([128, 3])
encoder.eo_patch_embed.S2_RGB.bias torch.Size([128])
encoder.eo_patch_embed.S2_Red_Edge.weight torch.Size([128, 3])
encoder.eo_patch_embed.S2_Red_Edge.bias torch.Size([128])
encoder.eo_patch_embed.S2_NIR_10m.weight torch.Size([128, 1])
encoder.eo_patch_embed.S2_NIR_10m.bias torch.Size([128])
encoder.eo_patch_embed.S2_NIR_20m.weight torch.Size([128, 1])
encoder.eo_patch_embed.S2_NIR_20m.bias torch.Size([128])
encoder.eo_patch_embed.S2_SWIR.weight torch.Size([128, 2])
encoder.eo_patch_embed.S2_SWIR.bias torch.Size([128])
encoder.eo_patch_embed.ERA5.weight torch.Size([128, 2])
encoder.eo_patch_embed.ERA5.bias torch.Size([128])
encoder.eo_patch_embed.SRTM.weight torch.Size([128, 2])
encoder.eo_patch_embed.SRTM.bias torch.Size([128])
encoder.eo_patch_embed.NDVI.weight torch.Size([128, 1])
encoder.eo_patch_embed.NDVI.bias torch.Size([128])
encoder.dw_embed.weight torch.Size([10, 128])
encoder.latlon_embed.weight torch.Size([128, 3])
encoder.latlon_embed.bias torch.Size([128])
encoder.blocks.0.norm1.weight torch.Size([128])
encoder.blocks.0.norm1.bias torch.Size([128])
encoder.blocks.0.attn.qkv.weight torch.Size([384, 128])
encoder.blocks.0.attn.qkv.bias torch.Size([384])
encoder.blocks.0.attn.proj.weight torch.Size([128, 128])
encoder.blocks.0.attn.proj.bias torch.Size([128])
encoder.blocks.0.norm2.weight torch.Size([128])
encoder.blocks.0.norm2.bias torch.Size([128])
encoder.blocks.0.mlp.fc1.weight torch.Size([512, 128])
encoder.blocks.0.mlp.fc1.bias torch.Size([512])
encoder.blocks.0.mlp.fc2.weight torch.Size([128, 512])
encoder.blocks.0.mlp.fc2.bias torch.Size([128])
encoder.blocks.1.norm1.weight torch.Size([128])
encoder.blocks.1.norm1.bias torch.Size([128])
encoder.blocks.1.attn.qkv.weight torch.Size([384, 128])
encoder.blocks.1.attn.qkv.bias torch.Size([384])
encoder.blocks.1.attn.proj.weight torch.Size([128, 128])
encoder.blocks.1.attn.proj.bias torch.Size([128])
encoder.blocks.1.norm2.weight torch.Size([128])
encoder.blocks.1.norm2.bias torch.Size([128])
encoder.blocks.1.mlp.fc1.weight torch.Size([512, 128])
encoder.blocks.1.mlp.fc1.bias torch.Size([512])
encoder.blocks.1.mlp.fc2.weight torch.Size([128, 512])
encoder.blocks.1.mlp.fc2.bias torch.Size([128])
encoder.norm.weight torch.Size([128])
encoder.norm.bias torch.Size([128])
encoder.month_embed.weight torch.Size([12, 32])
encoder.channel_embed.weight torch.Size([10, 32])
head.linear.weight torch.Size([10, 128])
head.linear.bias torch.Size([10])

**Comment by mahrokh3409** [comment]
Hi @gabrieltseng 
I used month=1 for the inference. I also tried fine-tuning the model with different numbers of samples to observe their impact on the inference. Although the fine-tuned models may predict different numbers, each number is unique for any image passed to the inference script. for example, the first fine-tuned model (100 samples per class) predicts number 1 for any image passed to the inference script. I also set ignore_dynamic_world = False 
I do appreciate your help.

Kind Regards,
Mahrokh
